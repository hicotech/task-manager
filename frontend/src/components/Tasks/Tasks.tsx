import type { Task, TaskList } from '@task-manager/shared-library';
import { useEffect, useState } from 'react';
import styled from 'styled-components';

export interface Props {
    /**
     * Unique ID for the component
     */
    id?: string;

    /**
     *  Tasks list name
     */
    name: TaskList['name'];

    /**
     * Task list items
     */
    tasks: TaskList['tasks'];

    /**
     *  On button click callback
     */
    onButtonClick?: (taskId: Task['id']) => void;

    /**
     * Allows to pass testid string for testing purposes
     */
    'data-testid'?: string;
}

const Container = styled.div`
    display: flex;
    flex-direction: column;
    gap: 0.5rem;
`;

const Title = styled.h1`
    text-align: center;
`;

const ProgressIndicator = styled.progress`
    width: 100%;
`;

const TasksList = styled.ul`
    display: flex;
    flex-direction: column;
    gap: 0.5rem;
    list-style: none;
    margin: 0;
    padding: 0;
`;

const Task = styled.li`
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 0.5rem;
    border-radius: 1rem;
    border: 1px dashed #335a6f;

    &:nth-child(even) {
        background-color: #e6e6e6;
    }
`;

const TaskButton = styled.button`
    display: inline-block;
    font-weight: 700;
    font-size: 1rem;
    line-height: 1.5rem;
    cursor: pointer;
    color: #528b7b;
    background-color: #ffffff;
    border-radius: 100rem;
    border: 2px solid #528b7b;
    padding: 0.375rem 1.25rem;

    &:hover {
        color: #ffffff;
        background-color: #528b7b;
    }

    &:active {
        border: 2px solid #ffffff;
        box-shadow: 0px 0px 0px 2px #528b7b;
    }
`;

const Tasks = ({ 'data-testid': dataTestId, ...props }: Props) => {
    const [initialListLength, setInitialListLength] = useState(0);

    // Store initial size of tasks array into local state
    useEffect(() => {
        setInitialListLength(props.tasks.length);
    }, []);

    return (
        <Container id={props.id} data-testid={dataTestId}>
            <Title>{props.name}</Title>
            <ProgressIndicator
                value={
                    initialListLength > 0
                        ? (initialListLength - props.tasks.length) / initialListLength
                        : initialListLength
                }
            />
            <TasksList>
                {props.tasks.map(task => (
                    <Task key={task.id}>
                        {task.name}
                        <TaskButton onClick={() => props.onButtonClick?.(task.id)}>
                            Complete Task
                        </TaskButton>
                    </Task>
                ))}
            </TasksList>
        </Container>
    );
};

/** @component */
export default Tasks;
