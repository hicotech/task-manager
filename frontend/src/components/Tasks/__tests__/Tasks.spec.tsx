import { fireEvent, render } from '@testing-library/react';
import type { Props as TasksProps } from '../Tasks';
import Tasks from '../Tasks';

describe('Tasks', () => {
    const testData: Required<TasksProps> = {
        id: 'tasks-id',
        name: 'tasks-name',
        tasks: [
            {
                id: 'task-1',
                name: 'Task Circle',
                type: 'circle',
            },
            {
                id: 'task-2',
                name: 'Task Rectangle',
                type: 'rectangle',
            },
            {
                id: 'task-3',
                name: 'Task Polygon',
                type: 'polygon',
            },
        ],
        onButtonClick: jest.fn(),
        'data-testid': 'tasks-testid',
    };

    describe('Renders Default', () => {
        it('Render list name', () => {
            const { container, getByText, getByRole } = render(
                <Tasks name={testData.name} tasks={testData.tasks} />
            );

            expect(container).toMatchSnapshot();

            expect(getByRole('heading')).toBeInTheDocument();
            expect(getByText(testData.name)).toBeInTheDocument();
        });

        it('Render progress bar', () => {
            const { getByRole } = render(<Tasks name={testData.name} tasks={testData.tasks} />);

            expect(getByRole('progressbar')).toBeInTheDocument();
        });

        it('Render three tasks', () => {
            const { container, getByText, getAllByRole, getByRole } = render(
                <Tasks name={testData.name} tasks={testData.tasks} />
            );

            expect(container).toMatchSnapshot();

            expect(getByRole('list')).toBeInTheDocument();
            expect(getAllByRole('listitem').length).toEqual(3);
            expect(getByText(testData.tasks[0].name)).toBeInTheDocument();
            expect(getByText(testData.tasks[1].name)).toBeInTheDocument();
            expect(getByText(testData.tasks[2].name)).toBeInTheDocument();
        });

        it('Render two tasks', () => {
            const { container, getByText, queryByText, getAllByRole, getByRole } = render(
                <Tasks name={testData.name} tasks={testData.tasks.slice(0, 2)} />
            );

            expect(container).toMatchSnapshot();

            expect(getByRole('list')).toBeInTheDocument();
            expect(getAllByRole('listitem').length).toEqual(2);
            expect(getByText(testData.tasks[0].name)).toBeInTheDocument();
            expect(getByText(testData.tasks[1].name)).toBeInTheDocument();
            expect(queryByText(testData.tasks[2].name)).not.toBeInTheDocument();
        });

        it('Render one task', () => {
            const { container, getByText, queryByText, getAllByRole, getByRole } = render(
                <Tasks name={testData.name} tasks={testData.tasks.slice(0, 1)} />
            );

            expect(container).toMatchSnapshot();

            expect(getByRole('list')).toBeInTheDocument();
            expect(getAllByRole('listitem').length).toEqual(1);
            expect(getByText(testData.tasks[0].name)).toBeInTheDocument();
            expect(queryByText(testData.tasks[1].name)).not.toBeInTheDocument();
            expect(queryByText(testData.tasks[2].name)).not.toBeInTheDocument();
        });
    });

    describe('Functionality testing', () => {
        it('Call onClick callback when button is clicked', () => {
            const handleClick = testData.onButtonClick;

            const { getByRole } = render(
                <Tasks
                    name={testData.name}
                    tasks={testData.tasks.slice(0, 1)}
                    onButtonClick={handleClick}
                />
            );

            fireEvent.click(getByRole('button'));
            expect(handleClick).toHaveBeenCalledTimes(1);
            expect(handleClick).toHaveBeenCalledWith(testData.tasks[0].id);
        });

        it('Call correct onClick callback when button is clicked', () => {
            const handleClick = jest.fn();

            const { getAllByRole } = render(
                <Tasks name={testData.name} tasks={testData.tasks} onButtonClick={handleClick} />
            );

            fireEvent.click(getAllByRole('button')[1]);
            expect(handleClick).toHaveBeenCalledTimes(1);
            expect(handleClick).toHaveBeenCalledWith(testData.tasks[1].id);
        });

        it('Update progress indicator correctly', () => {
            const { getByRole, rerender } = render(
                <Tasks name={testData.name} tasks={testData.tasks} />
            );

            expect(getByRole('progressbar')).toHaveAttribute('value', '0');

            // Simulate removal of Task from tasks list
            rerender(<Tasks name={testData.name} tasks={testData.tasks.slice(1)} />);
            expect(getByRole('progressbar')).toHaveAttribute('value', '0.3333333333333333');
        });
    });

    /**
     * Tests for name, tasks and onButtonClick are skipped as they would duplicate tests from "Functionality testing" section
     */
    describe('Props testing', () => {
        it('Id', () => {
            const { getByTestId } = render(
                <Tasks
                    name={testData.name}
                    tasks={testData.tasks}
                    id={testData.id}
                    data-testid={testData['data-testid']}
                />
            );

            expect(getByTestId(testData['data-testid'])).toHaveAttribute('id', testData.id);
        });

        it('Data-testid', () => {
            const { getByTestId } = render(
                <Tasks
                    name={testData.name}
                    tasks={testData.tasks}
                    data-testid={testData['data-testid']}
                />
            );

            expect(getByTestId(testData['data-testid'])).toBeInTheDocument();
        });
    });
});
