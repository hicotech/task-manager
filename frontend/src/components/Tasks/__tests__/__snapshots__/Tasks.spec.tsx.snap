// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks Renders Default Render list name 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

.c1 {
  text-align: center;
}

.c2 {
  width: 100%;
}

.c3 {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c4 {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0.5rem;
  border-radius: 1rem;
  border: 1px dashed #335a6f;
}

.c4:nth-child(even) {
  background-color: #e6e6e6;
}

.c5 {
  display: inline-block;
  font-weight: 700;
  font-size: 1rem;
  line-height: 1.5rem;
  cursor: pointer;
  color: #528b7b;
  background-color: #ffffff;
  border-radius: 100rem;
  border: 2px solid #528b7b;
  padding: 0.375rem 1.25rem;
}

.c5:hover {
  color: #ffffff;
  background-color: #528b7b;
}

.c5:active {
  border: 2px solid #ffffff;
  box-shadow: 0px 0px 0px 2px #528b7b;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      tasks-name
    </h1>
    <progress
      class="c2"
      value="0"
    />
    <ul
      class="c3"
    >
      <li
        class="c4"
      >
        Task Circle
        <button
          class="c5"
        >
          Complete Task
        </button>
      </li>
      <li
        class="c4"
      >
        Task Rectangle
        <button
          class="c5"
        >
          Complete Task
        </button>
      </li>
      <li
        class="c4"
      >
        Task Polygon
        <button
          class="c5"
        >
          Complete Task
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Tasks Renders Default Render one task 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

.c1 {
  text-align: center;
}

.c2 {
  width: 100%;
}

.c3 {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c4 {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0.5rem;
  border-radius: 1rem;
  border: 1px dashed #335a6f;
}

.c4:nth-child(even) {
  background-color: #e6e6e6;
}

.c5 {
  display: inline-block;
  font-weight: 700;
  font-size: 1rem;
  line-height: 1.5rem;
  cursor: pointer;
  color: #528b7b;
  background-color: #ffffff;
  border-radius: 100rem;
  border: 2px solid #528b7b;
  padding: 0.375rem 1.25rem;
}

.c5:hover {
  color: #ffffff;
  background-color: #528b7b;
}

.c5:active {
  border: 2px solid #ffffff;
  box-shadow: 0px 0px 0px 2px #528b7b;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      tasks-name
    </h1>
    <progress
      class="c2"
      value="0"
    />
    <ul
      class="c3"
    >
      <li
        class="c4"
      >
        Task Circle
        <button
          class="c5"
        >
          Complete Task
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Tasks Renders Default Render three tasks 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

.c1 {
  text-align: center;
}

.c2 {
  width: 100%;
}

.c3 {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c4 {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0.5rem;
  border-radius: 1rem;
  border: 1px dashed #335a6f;
}

.c4:nth-child(even) {
  background-color: #e6e6e6;
}

.c5 {
  display: inline-block;
  font-weight: 700;
  font-size: 1rem;
  line-height: 1.5rem;
  cursor: pointer;
  color: #528b7b;
  background-color: #ffffff;
  border-radius: 100rem;
  border: 2px solid #528b7b;
  padding: 0.375rem 1.25rem;
}

.c5:hover {
  color: #ffffff;
  background-color: #528b7b;
}

.c5:active {
  border: 2px solid #ffffff;
  box-shadow: 0px 0px 0px 2px #528b7b;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      tasks-name
    </h1>
    <progress
      class="c2"
      value="0"
    />
    <ul
      class="c3"
    >
      <li
        class="c4"
      >
        Task Circle
        <button
          class="c5"
        >
          Complete Task
        </button>
      </li>
      <li
        class="c4"
      >
        Task Rectangle
        <button
          class="c5"
        >
          Complete Task
        </button>
      </li>
      <li
        class="c4"
      >
        Task Polygon
        <button
          class="c5"
        >
          Complete Task
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Tasks Renders Default Render two tasks 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

.c1 {
  text-align: center;
}

.c2 {
  width: 100%;
}

.c3 {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c4 {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0.5rem;
  border-radius: 1rem;
  border: 1px dashed #335a6f;
}

.c4:nth-child(even) {
  background-color: #e6e6e6;
}

.c5 {
  display: inline-block;
  font-weight: 700;
  font-size: 1rem;
  line-height: 1.5rem;
  cursor: pointer;
  color: #528b7b;
  background-color: #ffffff;
  border-radius: 100rem;
  border: 2px solid #528b7b;
  padding: 0.375rem 1.25rem;
}

.c5:hover {
  color: #ffffff;
  background-color: #528b7b;
}

.c5:active {
  border: 2px solid #ffffff;
  box-shadow: 0px 0px 0px 2px #528b7b;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      tasks-name
    </h1>
    <progress
      class="c2"
      value="0"
    />
    <ul
      class="c3"
    >
      <li
        class="c4"
      >
        Task Circle
        <button
          class="c5"
        >
          Complete Task
        </button>
      </li>
      <li
        class="c4"
      >
        Task Rectangle
        <button
          class="c5"
        >
          Complete Task
        </button>
      </li>
    </ul>
  </div>
</div>
`;
